packages/@aws-cdk/pipelines/README.md

```diff
@@ -62,6 +107,524 @@ class MyApplication extends Stage {
+
+```ts
+new CodeBuildStep('Synth', {
+  // ...standard RunScript props...
+  commands: [/* ... */],
+  env: { /* ... */ },
+
+  // If you are using a CodeBuildStep explicitly, set the 'cdk.out' directory
+  // to be the synth step's output.
+  primaryOutputDirectory: 'cdk.out',
+
+  // Control the name of the project
+  projectName: 'MyProject',
+
+  // Control parts of the BuildSpec other than the regular 'build' and 'install' commands
+  partialBuildSpec: codebuild.BuildSpec.fromObject({
+    version: '0.2',
+    // ...
+  }),
+
+  // Control the build environment
+  buildEnvironment: {
+    computeType: codebuild.ComputeType.LARGE,
+  },
+
+  // Control Elastic Network Interface creation
+  vpc: vpc,
+  subnetSelection: { subnetType: ec2.SubnetType.PRIVATE },
+  securityGroups: [mySecurityGroup],
+
+  // Additional policy statements for the execution role
+  rolePolicy: [
+    new iam.PolicyStatement({ /* ... */ }),
+  ],
+});
+```
+
+You can also configure defaults for *all* CodeBuild projects by passing `codeBuildDefaults`,
+or just for the asset publishing and self-mutation projects by passing `assetPublishingCodeBuildDefaults`
+or `selfMutationCodeBuildDefaults`:
+
+```ts
+new CodePipeline(this, 'Pipeline', {
+  // ...
+
+  // Defaults for all CodeBuild projects
+  codeBuildDefaults: {
+    // Prepend commands and configuration to all projects
+    partialBuildSpec: codebuild.BuildSpec.fromObject({
+      version: '0.2',
+      // ...
+    }),
+
+    // Control the build environment
+    buildEnvironment: {
+      computeType: codebuild.ComputeType.LARGE,
+    },
+
+    // Control Elastic Network Interface creation
+    vpc: vpc,
+    subnetSelection: { subnetType: ec2.SubnetType.PRIVATE },
+    securityGroups: [mySecurityGroup],
+
+    // Additional policy statements for the execution role
+    rolePolicy: [
+      new iam.PolicyStatement({ /* ... */ }),
+    ],
+  },
+
+  assetPublishingCodeBuildDefaults: { /* ... */ },
+  selfMutationCodeBuildDefaults: { /* ... */ },
+});
+```
+
+### Building Docker image assets, and file asset bundling using Docker
+
+Docker can be used in 3 different places in the pipeline:
+
+* If you are using Docker image assets in your application stages: Docker will
+  run in the asset publishing projects.
+* If you are using Docker image assets in your stack (for example as
+  images for your CodeBuild projects): Docker will run in the self-mutate project.
+* If you are using Docker to bundle file assets anywhere in your project (for
+  example, if you are using such construct libraries as
+  `@aws-cdk/aws-lambda-nodejs`): Docker will run in the
+  *synth* project.
+
+For the first case, you don't need to do anything special. For the other two cases,
+you need to make sure that **privileged mode** is enabled on the correct CodeBuild
+projects, so that Docker can run correctly.
+
+If you are using asset bundling, use a `CodeBuildStep` as the `synth` step
```

Comment:
This is a really poor experience for people who use asset bundling. We have many users who love `NodeJsFunction` and friends and this is adding quite a lot of friction for them.

What can we do to improve this UX?

1. Is there something we can do to identify that a `Stage` uses asset bundling and auto-enable this for the synth step?
2. As a second-best option maybe we can add an option at the `CodePipelineProps` level (e.g. `dockerEnabledDuringSynth: true`.

Also - can you please separate this to two sections? One is about enabling asset bundling such as `NodejsFunction` and the other is about using a user-defined docker image asset as in CodeBuildStep.

Reply:
> Is there something we can do to identify that a Stage uses asset bundling and auto-enable this for the synth step?

Since asset bundling runs during Synth, by the time asset bundling runs it's already too late (you should have already enabled it).

But yeah, we can move the property top-level.